Reintroduce get_plugin_manager() for backward-compatibility

nicoddemus opened the pull request:
PyCharm depends on `get_plugin_manager()` for its pytest runner.

Using the pytest runner in PyCharm as of `2.8.0.dev4` raises this exception:

```
Traceback (most recent call last):
  File "D:\Programming\PyCharm\helpers\pycharm\pytestrunner.py", line 60, in <module>
    main()
  File "D:\Programming\PyCharm\helpers\pycharm\pytestrunner.py", line 30, in main
    _pluginmanager = get_plugin_manager()
  File "D:\Programming\PyCharm\helpers\pycharm\pytestrunner.py", line 20, in get_plugin_manager
    from _pytest.core import PluginManager
ImportError: No module named '_pytest.core'
```

The related code in `PyCharm\helpers\pycharm\pytestrunner.py` is:

```python
def get_plugin_manager():
  try:
    from _pytest.config import get_plugin_manager
    return get_plugin_manager()
  except ImportError:
    from _pytest.core import PluginManager
    return PluginManager(load=True)
```

`get_plugin_manager` was removed during the refactoring that lead to `pluggy`, I guess.

While it could be argued that PyCharm could fix it on their side, keeping backward compatibility in pytest is simple and would prevent a ton of PyCharm users not being able to run py.test tests when we release `2.8.0`, at least until a hotfix for PyCharm is released.

RonnyPfannschmidt commented:
nicely done
please add a deprecationwarning with stacklevel 2, so its noticeable more easily

hpk42 commented:
sure,a dding it back is fine but could you add a test noting pycharms needs it so the next person doesn't accidentally delete this method?

nicoddemus commented:
Applied suggested changes.

hpk42 commented on _pytest/config.py:
```diff
@@ -82,6 +82,12 @@ def get_config():
+    """deprecated, backward-compatibility for pytest < 2.8 (see #787)"""
+    warnings.warn('get_plugin_manager() is deprecated, use get_config().pluginmanager instead',
+                  DeprecationWarning, stacklevel=2)
+    return get_config().pluginmanager
```
sorry i didn't say this earlier but i think we don't need to bother with deprecating it.  @RonnyPfannschmidt , do you insist to deprecate it?  I think it's ok to offer a method for IDEs and others to get a plugin manager instance.

nicoddemus replied:
I agree we should just keep this around... waiting for @RonnyPfannschmidt's take on it.

RonnyPfannschmidt replied:
i'm wondering more about having a pytest-controller package,
that would support running py.test slave processes in various configurations

nicoddemus replied:
That's an interesting idea, but certainly out of the scope of this PR.

hpk42 replied:
Let's strike the deprecation and rather document get_plugin_manager() as official API (in plugins.txt i guess)

nicoddemus commented:
Removed the deprecation warning and added a reference to `get_plugin_manager` in `writing_plugins.txt`, as it looked like the better fit for this particular function.

While writing the docs for `get_plugin_manager` I realized there are no reference in the docs to `pluggy.PluginManager` or `PytestPluginManager`, so I added some to see how they would look like. My question is, should I continue to add more docs, or that API is not stable yet so it would be better to not even mentioning them at all, leaving only the reference to the `get_plugin_manager` function?

nicoddemus commented:
Decided to merge, we can decide on improving the documentation later. :smile:
